plankton/scripts/run.py

```diff
@@ -73,41 +71,34 @@ def do_run_simulation(argument_list=None):
     if arguments.add_path is not None:
         sys.path.append(os.path.abspath(arguments.add_path))
 
+    device_registry = DeviceRegistry(arguments.device_package)
+
     if not arguments.device:
-        devices = ['Please specify a device to simulate.',
-                   'The following devices are available:']
+        devices = ['Please specify a device to simulate. The following devices are available:']
 
-        for dev in get_available_submodules(arguments.device_package):
-            devices.append('\t' + dev)
+        for dev in device_registry.devices:
+            devices.append('    ' + dev)
 
         print('\n'.join(devices))
         return
 
-    # Import the device type and required initialisation parameters.
-    device_type, parameters = import_device(arguments.device, arguments.setup,
-                                            device_package=arguments.device_package)
+    device_builder = device_registry.device_builder(arguments.device)
 
     if arguments.list_protocols:
-        adapters = get_available_adapters(
-            arguments.device, device_package=arguments.device_package)
-
-        protocols = {adapter.protocol for adapter in adapters.values()}
-
-        print('\n'.join(protocols))
+        print('\n'.join(device_builder.protocols))
         return
 
-    device = device_type(**parameters)
-    adapter = import_adapter(
-        arguments.device, arguments.protocol,
-        device_package=arguments.device_package)(device, arguments.adapter_args)
+    device = device_builder.create_device(arguments.setup)
+    interface = device_builder.create_interface_type(arguments.protocol)(
+        device, arguments.adapter_args)
```

Comment:
What I kind of don't like with this approach is that it drags in this entire adapter command lines arguments thing into the factory. I'd be more comfortable with a `**kwargs` type solution to avoid duplicating the adapter signature here. I'll try and implement that.